### PR TITLE
Add Pipeline support && generic update

### PR DESCRIPTION
**Pull request by oleg-nenashev**
I'm working on a pet project, which involves automation for Jenkins Pipeline based on libs. I need to test libs without committing changes, hence I decided to use Filesystem SCM plugin for such purpose. The plugin was very outdated (last release - 2011), hence there is a bunch of changes

- [x] Upgrade to the new Parent POM
- [x] Upgrade to the new core baseline (1.642.x)
- [x] Rework of the UI to use DataBound constructor (see screenshot of the changes)
- [x] Fixes of issues reported by FindBugs/Functional tests
- [x] Update SCM API to support Jobs/Runs
- [x] Delete button for the excludes/includes list

After the changes the plugin supports the `scm` step and can be used as a Library SCM source. Functional tests will be added later if original maintainers +1 the approach (CC @kohsuke, @samngms and @kutzi )

Old Config UI:

screen shot 2017-01-01 at 20 42 24

New config UI:

screen shot 2017-01-02 at 00 18 31

@reviewbybees

### CODE REVIEW

**Comment by jglick** [comment]
> I need to test libs without committing changes, hence I decided to use Filesystem SCM plugin for such purpose.

Why not just use **Replay**?

**Comment by oleg-nenashev** [comment]
> Why not just use Replay?

How it can help with Libraries? From what I see I will have to commit/push them every time. Not thet comfortable for the local development.

**Comment by jglick** [comment]
>> Why not just use Replay?
>
> How it can help with Libraries? From what I see I will have to commit/push them every time.

No, at least for untrusted libraries you can replay library changes as well as main script or `load`ed script changes. For example, try

    make -C workflow-aggregator-plugin/demo run

**Comment by BerndPfaff** [comment]
I used this snapshot in a Jenkins pipeline job. At first it was working as expected.
After that I tried to change the configuration of this job in the UI. But the SCM part of the pipeline configuration was changed to a different scm tool (e.g. CVS). It was also not possible to select "File System" as SCM tool anymore, even some other options were not available anymore in the drop-down menu.
Only after the configuration Pipeline Definitions was saved as "Pipeline script", instead of "Pipeline script from SCM", it was possible to change the configuration back as "Pipeline script from SCM" File System.
Is this a known issue?

**Comment by oleg-nenashev** [comment]
@BerndPfaff Nope, it is not. Will try to reproduce once I find some time to finish this pull request

**Comment by oleg-nenashev** [comment]
Retriggering CI

**Comment by niothiel** [comment]
:+1: Would love to see this update. We are doing something very similar.

**Comment by oleg-nenashev** [comment]
I hope to finalize it soon

**Comment by oleg-nenashev** [comment]
@reviewbybees back to review.

File: pom.xml [hunk 1]
```diff
@@ -31,6 +35,14 @@
     </developer>
   </developers>
   
+  <licenses>
+    <license>
+      <name>MIT License</name>
+      <url>http://www.opensource.org/licenses/mit-license.php</url>
```
Comment (jglick) [inline pom.xml]: or just https://opensource.org/licenses/MIT

File: pom.xml [hunk 2]
```diff
@@ -46,7 +58,33 @@
             <url>http://repo.jenkins-ci.org/public/</url>
         </pluginRepository>
     </pluginRepositories>
+    
     <dependencies>
+      <dependency>
+        <groupId>org.jenkins-ci.plugins.workflow</groupId>
+        <artifactId>workflow-aggregator</artifactId>
```
Comment (jglick) [inline pom.xml]: Do not use this plugin as a dependency even for tests. Specify `workflow-cps`, `workflow-job`, and any step plugins you need.

File: src/test/java/hudson/plugins/filesystem_scm/integration/pipeline/PipelineLibraryTest.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
```
Comment (jglick) [inline src/test/java/hudson/plugins/filesystem_scm/integration/pipeline/PipelineLibraryTest.java]: fix

File: src/test/java/hudson/plugins/filesystem_scm/integration/pipeline/PipelineLibraryTest.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+ */
+package hudson.plugins.filesystem_scm.integration.pipeline;
+
+/**
+ *
+ * @author Oleg Nenashev
+ */
+public class PipelineLibraryTest {
+    
+}
```
Comment (jglick) [inline src/test/java/hudson/plugins/filesystem_scm/integration/pipeline/PipelineLibraryTest.java]: ???

File: pom.xml [hunk 5]
```diff
@@ -46,7 +58,33 @@
...
     </pluginRepositories>
+    
     <dependencies>
+      <dependency>
+        <groupId>org.jenkins-ci.plugins.workflow</groupId>
+        <artifactId>workflow-aggregator</artifactId>
+        <version>2.4</version>
+        <scope>test</scope>
+      </dependency>
+      <!-- Transient upper bound dependencies -->
```
Comment (jglick) [inline pom.xml]: Transitive?

File: src/main/java/hudson/plugins/filesystem_scm/FSSCM.java [hunk 6]
```diff
@@ -38,60 +41,88 @@
...
+    /**
+     * Filter settings.
+     *
+     * @since TODO
+     */
+    @CheckForNull
+    private FilterSettings filterSettings;
+
+    @DataBoundConstructor
+    public FSSCM(String path, boolean clearWorkspace, boolean copyHidden, FilterSettings filterSettings) {
```
Comment (jglick) [inline src/main/java/hudson/plugins/filesystem_scm/FSSCM.java]: Use `@DataBoundSetter`.

File: pom.xml [hunk 7]
```diff
@@ -46,7 +58,33 @@
...
+      <dependency>
+        <groupId>org.jenkins-ci.plugins</groupId>
+        <artifactId>scm-api</artifactId>
+        <version>1.3</version>
+        <scope>test</scope>
+      </dependency>
+      <dependency>
+        <groupId>org.jenkins-ci.plugins</groupId>
+        <artifactId>structs</artifactId>
```
Comment (jglick) [inline pom.xml]: Make a `compile`-time dependency and add `@Symbol`s.

File: src/main/java/hudson/plugins/filesystem_scm/FolderDiff.java [hunk 8]
```diff
@@ -1,25 +1,27 @@
...
  *   <li>check if there are deleted files in the source folder</li>
  * </ul>
- * </p>
- * 
+ * @param <T> Type of the item being returned by the callable
  * @author Sam NG
  *
  */
-public class FolderDiff implements Serializable {
+public class FolderDiff<T> extends SlaveToMasterFileCallable<T> implements Serializable {
```
Comment (jglick) [inline src/main/java/hudson/plugins/filesystem_scm/FolderDiff.java]: No! This is a `MasterToSlaveFileCallable`.

File: src/main/java/hudson/plugins/filesystem_scm/RemoteCopyDir.java [hunk 9]
```diff
@@ -5,10 +5,10 @@
 
 import org.apache.commons.io.FileUtils;
 
-import hudson.FilePath.FileCallable;
 import hudson.remoting.VirtualChannel;
+import jenkins.SlaveToMasterFileCallable;
 
-public class RemoteCopyDir implements FileCallable<Boolean> {
+public class RemoteCopyDir extends SlaveToMasterFileCallable<Boolean> {
```
Comment (jglick) [inline src/main/java/hudson/plugins/filesystem_scm/RemoteCopyDir.java]: Probably also wrong, though the class appears unused anyway so maybe just delete it.

File: src/main/java/hudson/plugins/filesystem_scm/RemoteFolderDiff.java [hunk 10]
```diff
@@ -8,8 +8,9 @@
 import java.io.*;
 import java.util.ArrayList;
 import java.util.List;
+import jenkins.security.SlaveToMasterCallable;
```
Comment (jglick) [inline src/main/java/hudson/plugins/filesystem_scm/RemoteFolderDiff.java]: unused import

File: src/main/resources/hudson/plugins/filesystem_scm/FSSCM/config.jelly [hunk 11]
```diff
@@ -1,46 +1,13 @@
+<?jelly escape-by-default='true'?>
 <j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form">
-  <f:entry title="Path" help="/plugin/filesystem_scm/help-path.html">
-    <f:textbox name="fs_scm.path" value="${scm.path}" />
+  <f:entry title="Path" field="path" help="/plugin/filesystem_scm/help-path.html">
```
Comment (jglick) [inline src/main/resources/hudson/plugins/filesystem_scm/FSSCM/config.jelly]: Move help file to this folder and `help` need not be specified.

File: src/main/resources/hudson/plugins/filesystem_scm/FSSCM/config.jelly [hunk 12]
```diff
@@ -1,46 +1,13 @@
...
+    <f:textbox/>
   </f:entry>
-  <f:entry title="Clear Workspace" help="/plugin/filesystem_scm/help-clearWorkspace.html">
-    <f:checkbox name="fs_scm.clearWorkspace" checked="${scm.clearWorkspace}"/>
+  <f:entry title="Clear Workspace" field="clearWorkspace" help="/plugin/filesystem_scm/help-clearWorkspace.html">
```
Comment (jglick) [inline src/main/resources/hudson/plugins/filesystem_scm/FSSCM/config.jelly]: Ditto.

File: src/main/resources/hudson/plugins/filesystem_scm/FSSCM/config.jelly [hunk 13]
```diff
@@ -1,46 +1,13 @@
...
-                    </f:entry>
-                  </table>
-                </f:repeatable>
-              </td>
-            </tr>
-          </table>
-        </f:block>
-      </f:optionalBlock>
-    </table>
+  <f:entry title="Copy Hidden Files/Folders" field="copyHidden" help="/plugin/filesystem_scm/help-copyHidden.html">
```
Comment (jglick) [inline src/main/resources/hudson/plugins/filesystem_scm/FSSCM/config.jelly]: Ditto.

File: pom.xml (hunk 1 above)
Comment (oleg-nenashev) [inline pom.xml]: Does Netbeans already autoresolve it?

File: src/main/resources/hudson/plugins/filesystem_scm/FSSCM/config.jelly (hunk 11 above)
Comment (oleg-nenashev) [inline src/main/resources/hudson/plugins/filesystem_scm/FSSCM/config.jelly]: It came from the old structure. I would prefer to refactor it later

File: pom.xml (hunk 1 above)
Comment (jglick) [inline pom.xml]: Yes I think so.

File: src/main/java/hudson/plugins/filesystem_scm/FSSCM.java [hunk 14]
```diff
@@ -303,6 +304,7 @@ private static String formatDuration(long diff) {
 	}
 
     @Extension
+    @Symbol("fsscm")
```
Comment (jglick) [inline src/main/java/hudson/plugins/filesystem_scm/FSSCM.java]: Should not include the `scm`. Maybe just `filesystem`?

File: src/main/java/hudson/plugins/filesystem_scm/FSSCM.java [hunk 15]
```diff
@@ -348,5 +350,12 @@ public PollingResult compareRemoteRevisionWith(Job<?, ?> project, Launcher launc
             return PollingResult.NO_CHANGES;
         }
     }
+
+    @Override
+    public boolean requiresWorkspaceForPolling() {
+        return super.requiresWorkspaceForPolling(); //To change body of generated methods, choose Tools | Templates.
```
Comment (jglick) [inline src/main/java/hudson/plugins/filesystem_scm/FSSCM.java]: Why?

File: src/main/java/hudson/plugins/filesystem_scm/RemoteCopyDir.java [hunk 16]
```diff
@@ -1,14 +1,23 @@
...
 
 import hudson.remoting.VirtualChannel;
-import jenkins.SlaveToMasterFileCallable;
+import jenkins.MasterToSlaveFileCallable;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
 
-public class RemoteCopyDir extends SlaveToMasterFileCallable<Boolean> {
+/**
+ * @deprecated Not used anymore
```
Comment (jglick) [inline src/main/java/hudson/plugins/filesystem_scm/RemoteCopyDir.java]: So just delete it.

File: src/test/java/hudson/plugins/filesystem_scm/integration/pipeline/PipelineLibraryTest.java [hunk 17]
```diff
@@ -1,14 +1,45 @@
...
+   // @Rule
+    //public TemporaryFolder tmpDir = new TemporaryFolder();
+    
+    @Test
+    public void shouldSupportFSSCMsJenkinsfileSource() throws Exception {
+        
+        // Init repo
+     //   File fsscmDir = tmpDir.newFolder("fsscm");
+    //    File jenkinsfile = new File(fsscmDir, "Jenkinsfile");
+    //    FileUtils.write(jenkinsfile, "echo `Hello, world!`");
```
Comment (jglick) [inline src/test/java/hudson/plugins/filesystem_scm/integration/pipeline/PipelineLibraryTest.java]: Why commented out?

Comment (oleg-nenashev) [inline src/test/java/hudson/plugins/filesystem_scm/integration/pipeline/PipelineLibraryTest.java]: The test is still the mess

File: src/main/java/hudson/plugins/filesystem_scm/RemoteCopyDir.java (hunk 16 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/filesystem_scm/RemoteCopyDir.java]: Breaks binary compatibility. I do not want to touch it for now

**Comment by oleg-nenashev** [comment]
OK, now the test just hangs

**Comment by oleg-nenashev** [comment]
@jglick @niothiel @BerndPfaff  I have released the current state as alpha-1. It should become available in the experimental UC soon. I will test it more before the release, especially the case from @BerndPfaff though I was unable to reproduce it.

**Comment by tanandy** [comment]
i got the same problem as @BerndPfaff

**Comment by oleg-nenashev** [comment]
@tanandy Ack, I will double check it once I get back to the plugin

**Comment by oleg-nenashev** [comment]
@tanandy @BerndPfaff Fixed that stupid issue in the code and released alpha-2. Sorry, was working on other things and this slipped down the priority list

**Comment by tanandy** [comment]
@oleg-nenashev thanks a lot :+1:

**Comment by maloups** [comment]
It works perfectly! :)
Thank you very much!

**Comment by oleg-nenashev** [comment]
I asked @jglick, he is fine with merging. I am about releasing it tomorrow. CC @abayer @svanoort just in case they have any comments/concerns.

**Comment by svanoort** [comment]
I probably won't be able to review before then.

**Comment by oleg-nenashev** [comment]
@svanoort just let me know if you **want** to review it. I can wait with release till Friday EoBD in your timezone. You know, conference-driven development :)

**Comment by oleg-nenashev** [comment]
🚢 🇮🇹

**Comment by oleg-nenashev** [comment]
I will not bump the version to 2.0, there is no significant compat breakages

**Comment by oleg-nenashev** [comment]
Well, I changed my mind after some consideration. In the worst case I will introduce 3.0 if I ever implement the SCM API

**Comment by oleg-nenashev** [comment]
Oh well, better to be on the safe side OTOH

**Pull request merged**